cloudmarker/test/test_filestore.py

```diff
@@ -0,0 +1,69 @@
+"""Tests for filestore plugins."""
+
+
+import os
+import pathlib
+import shutil
+import unittest
+
+from cloudmarker.stores import filestore
+
+
+class FileStoreTest(unittest.TestCase):
+    """Tests for filestore plugins."""
+
+    @classmethod
+    def write_to_file(cls, filename):
+        # Initialize the FileStore plugin with the filename
+        store = filestore.FileStore(filename)
+
+        # dummy data to write
+        mp = {
+            "record_type": "dummy",
+            "data": {
+                "test_name": "tilde_filename"
+            }
+        }
+
+        store.write(mp)
+        store.done()
+
+    def test_filepath_tilde(self):
+        self.write_to_file("~/caramel_l33t/cloudmarker")
+        # if the parsing fails, it should create a "~" directory in the current
+        # path
+        f = pathlib.Path("~")
+
+        # If exists and is a directory, the parsing failed.
+        self.assertFalse(f.exists() and f.is_dir())
+
+    def test_filepath_tilde2(self):
+        user = os.getenv("USER")
+        self.write_to_file("~{}/caramel_tpb/cloudmarker".format(user))
+        # if the parsing fails, it should create a "~" directory in the current
+        # path
+        f = pathlib.Path("~{}".format(user))
+
+        # If exists and is a directory, the parsing failed.
+        self.assertFalse(f.exists() and f.is_dir())
+
+    def test_filepath_dir_traversal(self):
+        self.write_to_file("/tmp/cloudmarker/../abc")
+
+        f = pathlib.Path("/tmp/abc")
+        self.assertTrue(f.exists() and f.is_dir())
+
+    @classmethod
+    def tearDownClass(cls):
```

Comment:
This should be an instance method, i.e., `tearDown(self)`.

Rationale: Every test should run in a clean environment. There should be no coupling between the tests. A `tearDown(self)` call after test method is run reduces the chance of decoupling.

When we see a `tearDownClass(cls)` class method, the reader of the test code has to spend some additional time reading the code of each test to ensure that one method does not inadvertently depend on artifacts left by another test. But if we see a `tearDown(self)` method instead, we know for sure that the environment is getting cleaned up after every test method runs, so if a test method inadvertently depends on artifacts left by another test, then that test method is likely to fail because a `tearDown(self)` call after every test would blow away those artifacts.

Reply:
The [`tearDownClass`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.tearDownClass) method is related to the `unittest` class and is supposed to be called after the completion of all the tests in that particular class. The library docs say that it should be marked as a `@classmethod`. If I used `self`, the linter says:
```
cloudmarker/test/test_filestore.py:62:4: C0202 Class method tearDownClass should have 'cls' as first argument [pylint]
```
The reason for not using `teadDown(cls)` is that it runs the same thing after every test. Since every test will generate a different directory, it seems to me that there should be one tearDown at the end which is `tearDownClass`.

Reply:
The `tearDownClass(cls)` method should be written with `@classmethod` and `cls` because it is a classmethod. This is alright.

I am suggesting that we do not use `tearDownClass(cls)` at all unless there is a very good reason to do so and we use `tearDown(self)` instead because it is run after every test. That way, we know that every test runs in a clean environment (to the extent possible). I offered decoupling of tests as the rationale behind this.

By decoupling of tests, I mean, we should not create a situation where it is possible for `test_method1(self)` to create an artifact that `test_method2(self)` uses. If `test_method2(self)` requires some artifacts for itself to function, then it should itself create those artifacts or they should be created just before the test method runs via `setUp(self)` (not to be confused with `setUpClass(self)`).

If we perform a tear down only at the end of all tests in a test class, then we might inadvertently (not now but sometime in future) create a file in `test_method1(self)` that `test_method2(self)` uses and we would not know about it if we are using `tearDownClass(cls)`. The tests would run fine and we would not know that we have coupled the two test methods. If, however, we use `tearDown(self)`, then `test_method2(self)` would fail because the artifacts it is inadvertently depending on would have been removed by `tearDown(self)`. After seeing this failure, we would know that we need to fix `test_method2(self)` and decouple it from `test_method1(self)`.

Having said that, you should address this comment only if you agree with the rationale. It is okay not to address this comment as well. We can merge this pull request with or without addressing this review comment because the test code is correct in either case.